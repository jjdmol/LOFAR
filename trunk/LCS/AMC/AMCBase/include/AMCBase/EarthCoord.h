//# EarthCoord.h: Class to hold an earth coordinate as lon,lat,height
//#
//# Copyright (C) 2002
//# ASTRON (Netherlands Foundation for Research in Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//# This program is free software; you can redistribute it and/or modify
//# it under the terms of the GNU General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or
//# (at your option) any later version.
//#
//# This program is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License
//# along with this program; if not, write to the Free Software
//# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//# $Id$

#ifndef LOFAR_AMCBASE_EARTHCOORD_H
#define LOFAR_AMCBASE_EARTHCOORD_H

// \file EarthCoord.h
// Class to hold an earth coordinate as lon,lat,height

//# Forward Declarations.
#include <Common/lofar_iosfwd.h>
#include <Common/lofar_string.h>

namespace LOFAR
{
  namespace AMC
  {

    // \addtogroup AMCBase
    // @{

    // This class represents a position on earth. The position is stored using
    // longitude, latitude and height. The context where the object is used
    // defines the coordinate system and frame, so the class can be used in
    // any kind of frame (like ITRF and geocentric). The correct
    // interpretation of the coordinates should be done by the user of this
    // class.
    class EarthCoord
    {
    public:
      // Types of earth coordinates. 
      enum Types {
        INVALID = -1,   ///< Used when specified value is out of range.
        ITRF,
        WGS84,
        //# Insert new types HERE !!
        N_Types         ///< Number of reference types.
      };

      // Default constructor uses 0 for the values.
      EarthCoord()
        : itsLong(0), itsLat(0), itsHeight(0), itsType(ITRF) {}

      // Create a earth coordinate by giving the longitude and latitude in
      // radians and the height in meters. Reference type can be either ITRF
      // (default), or WGS84.
      EarthCoord (double longitude, double latitude, double height=0,
                  Types typ = ITRF);

      // Return the longitude in radians.
      double longitude() const
      { return itsLong; }

      // Return the latitude in radians.
      double latitude() const
      { return itsLat; }

      // Return the height in meters.
      double height() const
      { return itsHeight; }

      // Return the reference type.
      Types type() const
      { return itsType; }

      // Return the reference type as a string.
      const string& showType() const;

      // Return whether sky coordinate type is valid.
      bool isValid() const
      { return itsType != INVALID; }

    private:
      // Longitude in radians.
      double itsLong;

      // Latitude in radians.
      double itsLat;

      // Height in meters.
      double itsHeight;

      // Type of earth coordinate.
      Types itsType;
    };

    // Output an EarthCoord in ASCII format.
    ostream& operator<< (ostream&, const EarthCoord&);

    // Compare two EarthCoord objects for equality.
    // \note Two invalid objects can \e never be equal.
    bool operator==(const EarthCoord& lhs, const EarthCoord& rhs);

    // @}

  } // namespace AMC

} // namespace LOFAR

#endif
