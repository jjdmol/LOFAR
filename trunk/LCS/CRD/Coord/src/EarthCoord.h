//# EarthCoord.h: Class to hold an earth coordinate as lon,lat,height
//#
//# Copyright (C) 2002
//# ASTRON (Netherlands Foundation for Research in Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//# This program is free software; you can redistribute it and/or modify
//# it under the terms of the GNU General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or
//# (at your option) any later version.
//#
//# This program is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License
//# along with this program; if not, write to the Free Software
//# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//# $Id$

#if !defined(COORD_EARTHCOORD_H)
#define COORD_EARTHCOORD_H

//# Forward Declarations.
#include <Common/lofar_iosfwd.h>

class EarthCoord
{
public:
  // Default constructor uses 0 for the values.
  EarthCoord()
    : itsLong(0), itsLat(0), itsHeight(0) {}

  // Create a earth coordinate by giving the longitude and latitude in radians
  // and the height in meters.
  EarthCoord (double longitude, double latitude, double height=0)
    : itsLong(longitude), itsLat(latitude), itsHeight(height) {}

  // Get the values out.
  double longitude() const
    { return itsLong; }
  double latitude() const
    { return itsLat; }
  double height() const
    { return itsHeight; }

  // Output in ASCII.
  friend ostream& operator<< (ostream&, const EarthCoord&);

private:
  double itsLong;
  double itsLat;
  double itsHeight;
};


#endif
