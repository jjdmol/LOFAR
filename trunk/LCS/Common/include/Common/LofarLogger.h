//#  LofarLogger.h: Macro interface to the lofar logging package
//#
//#  Copyright (C) 2004
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

#if !defined(COMMON_LOFAR_LOGGER_H)
#define COMMON_LOFAR_LOGGER_H

#include <lofar_config.h>
#include <Common/StringUtil.h>

//# Dependent of the availability of the log4Cplus package one set of macro's
//# or the other set of macro's is included.
#ifdef HAVE_LOG4CPLUS
#include <Common/LofarLog4Cplus.h>
#else
#include <Common/LofarLogCout.h>
#endif //HAVE_LOG4CPLUS

namespace LOFAR 
{
  // Define the LOFAR::Exception class.
  EXCEPTION_CLASS(AssertError,Exception)  
}

//# --------------------- Common definitions ---------------

//# Make sure that LOFARLOGGER_PACKAGE is defined.
#ifndef LOFARLOGGER_PACKAGE
# define LOFARLOGGER_PACKAGE "Unknown_package"
#endif

//#
//# AUTO_FUNCTION_NAME
//#
// The macro AUTO_FUNCTION_NAME will be resolved by the (pre)compiler to 
// hold the name of the function the macro was used in.
#if defined(HAVE_PRETTY_FUNCTION)
#	define AUTO_FUNCTION_NAME		__PRETTY_FUNCTION__
#elif defined(HAVE_FUNCTION)
#	define AUTO_FUNCTION_NAME		__FUNCTION__
#else
#	define AUTO_FUNCTION_NAME		"??"
#endif

// NB: THROW is defined in package-dependant file.

// If the condition of assert is NOT met a logrequest is sent to the logger
// <module>.EXCEPTION and an AsserError exception is thrown.
#define ASSERT(cond) \
	if (!(cond))  THROW(LOFAR::AssertError, "Assertion: " #cond)
#define ASSERTSTR(cond,stream) \
	if (!(cond))  THROW(LOFAR::AssertError, "Assertion: " #cond "; " << stream)

// If the condition of failwhen is met a logrequest is sent to the logger
// <module>.EXCEPTION and an AsserError exception is thrown.
#define FAILWHEN(cond) \
	if (cond)  THROW(LOFAR::AssertError, "Failtest: " #cond)

// The DBG... version of ASSERT and FAILWHEN will only be in your compiled
// code when the (pre)compiler flag ENABLE_DBGASSERT is defined.
#ifdef ENABLE_DBGASSERT
#define DBGASSERT(cond)				ASSERT(cond)
#define DBGASSERTSTR(cond,stream)	ASSERTSTR(cond,stream)
#define DBGFAILWHEN(cond)			FAILWHEN(cond)
#else
#define DBGASSERT(cond)
#define DBGASSERTSTR(cond,stream)
#define DBGFAILWHEN(cond)
#endif

#endif // file read before
