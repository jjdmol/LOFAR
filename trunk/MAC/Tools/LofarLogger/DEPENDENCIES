Compiler version:
compat-2.96 (special compiler for the RedHat 8.0 distribution; almost the same as 2.95 but extended with STL)
Because the 3.2 is already installed with RedHat 8.0 the gcc and g++ of the compat 
installation must be renamed (to for instance gcc296 and g++296).

EXTERNAL PACKAGES:
		
log4cplus-1.0
	Prepare build:
	> CC=gcc296 CXX=g++296 CXXFLAGS="-g -O2 -fno-rtti" sh ./configure --enable-threads=no
	- set '#define HAVE_SSTREAM 1' to comment
	- ignore compiler erros during compiling the 'test''s
	- ignore compiler error during installing the 'test's

LCS/Common
