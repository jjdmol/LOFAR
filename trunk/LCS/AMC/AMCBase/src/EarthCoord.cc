//# EarthCoord.cc: Class to hold an earth coordinate as lon,lat,height
//#
//# Copyright (C) 2002
//# ASTRON (Netherlands Foundation for Research in Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//# This program is free software; you can redistribute it and/or modify
//# it under the terms of the GNU General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or
//# (at your option) any later version.
//#
//# This program is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License
//# along with this program; if not, write to the Free Software
//# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//# $Id$

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

//# Includes
#include <AMCBase/EarthCoord.h>
#include <Common/lofar_iostream.h>
#include <cmath>

namespace LOFAR
{
  namespace AMC
  {

    EarthCoord::EarthCoord (double longitude, double latitude, double height,
                            Types typ)
      : itsLong(longitude), itsLat(latitude), itsHeight(height)
    {
      if (INVALID < typ && typ < N_Types) itsType = typ; 
      else itsType = INVALID;
    }


    const string& EarthCoord::showType() const
    {
      //# Caution: Always keep this array of strings in sync with the enum
      //#          Types that is defined in the header file!
      static const string types[N_Types+1] = {
        "ITRF",
        "WGS84",
        "<INVALID>"
      };
      if (isValid()) return types[itsType];
      else return types[N_Types];
    }


    vector<double> EarthCoord::xyz() const
    {
      vector<double> p(3);
      double tmp = std::cos(itsLat);
      p[0] = itsHeight * std::cos(itsLong) * tmp;
      p[1] = itsHeight * std::sin(itsLong) * tmp;
      p[2] = itsHeight * std::sin(itsLat);
      return p;
    }


    ostream& operator<<(ostream& os, const EarthCoord& pos)
    {
      if (!pos.isValid()) 
        return os << pos.showType();
      else 
        return os << "["  << pos.longitude() << ", " << pos.latitude() 
                  << ", " << pos.height() << "] (" << pos.showType() << ")";
    }


    bool operator==(const EarthCoord& lhs, const EarthCoord& rhs)
    {
      return 
        lhs.isValid()   && rhs.isValid()   &&
        lhs.longitude() == rhs.longitude() &&
        lhs.latitude()  == rhs.latitude()  &&
        lhs.height()    == rhs.height()    &&
        lhs.type()      == rhs.type();
    }

  } // namespace AMC

} // namespace LOFAR
