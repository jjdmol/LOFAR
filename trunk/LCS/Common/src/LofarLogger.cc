//#  LofarLogger.cc: Interface to the log4cplus logging package
//#
//#  Copyright (C) 2004
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

//# Includes
#include <Common/LofarLogger.h>

namespace LOFAR {

//#------------------------- Global user functions -------------------------------
//
// formatString(format, ...) --> string
//
// Define a global function the accepts printf like arguments and returns a string.
//
const std::string formatString(const	char* format, ...) {
	char		tmp_cstring[1024];
	va_list		ap;

	va_start (ap, format);
	vsnprintf(tmp_cstring, sizeof(tmp_cstring), format, ap);
	va_end   (ap);

	return   std::string(tmp_cstring);
}


//#------------------------- Internal implementation -----------------------------
using namespace log4cplus;

//# ------------------------ implement the five trace levels ------------------------
const LogLevel	TRACE1_LOG_LEVEL	= 1;
const LogLevel	TRACE2_LOG_LEVEL	= 2;
const LogLevel	TRACE3_LOG_LEVEL	= 3;
const LogLevel	TRACE4_LOG_LEVEL	= 4;
const LogLevel	TRACE5_LOG_LEVEL	= 5;

#define _TRACE1_STRING	LOG4CPLUS_TEXT("TRACE1")
#define _TRACE2_STRING	LOG4CPLUS_TEXT("TRACE2")
#define _TRACE3_STRING	LOG4CPLUS_TEXT("TRACE3")
#define _TRACE4_STRING	LOG4CPLUS_TEXT("TRACE4")
#define _TRACE5_STRING	LOG4CPLUS_TEXT("TRACE5")

tstring traceLevel2String(LogLevel ll) {
	switch (ll) {
	case TRACE1_LOG_LEVEL:	return _TRACE1_STRING;
	case TRACE2_LOG_LEVEL:	return _TRACE2_STRING;
	case TRACE3_LOG_LEVEL:	return _TRACE3_STRING;
	case TRACE4_LOG_LEVEL:	return _TRACE4_STRING;
	case TRACE5_LOG_LEVEL:	return _TRACE5_STRING;
	}

	return tstring();		// not found
}

LogLevel string2TraceLevel (const tstring& lname) {
	if (lname == _TRACE1_STRING)	return TRACE1_LOG_LEVEL;
	if (lname == _TRACE2_STRING)	return TRACE2_LOG_LEVEL;
	if (lname == _TRACE3_STRING)	return TRACE3_LOG_LEVEL;
	if (lname == _TRACE4_STRING)	return TRACE4_LOG_LEVEL;
	if (lname == _TRACE5_STRING)	return TRACE5_LOG_LEVEL;

	return NOT_SET_LOG_LEVEL;			// not found
}

// initTracemodule
//
// Function that is used when the TRACE levels are NOT compiled out. It registers
// the TRACEn management routines at the Log4Cplus LogLevelManager and sets up the
// global trace-logger named "TRACE", with the additivity set to false.
// Attached to the trace-logger is one Appender that logs to stderr.
//
void initTraceModule (void) {
	//# register our own loglevels
	getLogLevelManager().pushToStringMethod(traceLevel2String);
	getLogLevelManager().pushFromStringMethod(string2TraceLevel);
	
	//# Setup a property object to initialise the TRACE Logger
	helpers::Properties		traceProp;		
	traceProp.setProperty("log4cplus.logger.TRACE", "TRACE, TRACE");
	traceProp.setProperty("log4cplus.additivity.TRACE", "false");
	traceProp.setProperty("log4cplus.appender.TRACE", "log4cplus::ConsoleAppender");
//	traceProp.setProperty("log4cplus.appender.TRACE.layout","log4cplus::PatternLayout");
//	traceProp.setProperty("log4cplus.appender.TRACE.layout.ConversionPattern","%c %m%n");
	traceProp.setProperty("log4cplus.appender.TRACE.layout", "log4cplus::TTCCLayout");
	traceProp.setProperty("log4cplus.appender.TRACE.layout.Use_gmtime", "true");

	PropertyConfigurator(traceProp).configure();

}
LOFAR::LoggerReference	TraceLoggerRef("TRACE");			// create the tracelogger

}	// namespace LOFAR
