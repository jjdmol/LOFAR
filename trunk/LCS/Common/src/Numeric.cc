//# Numeric.cc: 
//#
//# Copyright (C) 2002-2004
//# ASTRON (Netherlands Institute for Radio Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands
//#
//# This file is part of the LOFAR software suite.
//# The LOFAR software suite is free software: you can redistribute it and/or
//# modify it under the terms of the GNU General Public License as published
//# by the Free Software Foundation, either version 3 of the License, or
//# (at your option) any later version.
//#
//# The LOFAR software suite is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with the LOFAR software suite. If not, see <http://www.gnu.org/licenses/>.
//#
//# $Id$

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

//# Includes
#include <Common/Numeric.h>
#include <Common/LofarLogger.h>
#include <cstdlib>

namespace LOFAR
{
  bool Numeric::isFinite(float f)
  {
    floatUnion mask = { f };
    return mask.mask & floatExponentMask != floatExponentMask;
  }

  bool Numeric::isFinite(double d)
  {
    doubleUnion mask = { d };
    return mask.mask & doubleExponentMask != doubleExponentMask;
  }

  bool Numeric::isNegative(float f)
  {
    floatUnion mask = { f };
    return mask.mask & floatNegativeMask == floatNegativeMask;
  }

  bool Numeric::isNegative(double d)
  {
    doubleUnion mask = { d };
    return mask.mask & doubleNegativeMask == doubleNegativeMask;
  }

  bool Numeric::isInf(float f)
  {
    floatUnion mask = { f };
    return !isFinite(f) && (mask.mask & floatMantissaMask == 0L);
  }

  bool Numeric::isInf(double d)
  {
    doubleUnion mask = { d };
    return !isFinite(d) && (mask.mask & doubleMantissaMask == 0LL);
  }

  bool Numeric::isNan(float f)
  {
    floatUnion mask = { f };
    return !isFinite(f) && (mask.mask & floatMantissaMask != 0L);
  }

  bool Numeric::isNan(double d)
  {
    doubleUnion mask = { d };
    return !isFinite(d) && (mask.mask & doubleMantissaMask != 0LL);
  }


  bool Numeric::compare(float lhs, float rhs, floatMask_t maxUlps)
  {

#ifdef USE_INFINITYCHECK
    // If either \a lhs or \a rhs is infinite, then return true if they're
    // exactly equal, i.e. they have the same sign.
    if (isInf(lhs) || isInf(rhs)) return lhs == rhs;
#endif

#ifdef USE_NANCHECK
    // Consider \a lhs and \a rhs unequal when either is NaN. 
    if (isNan(lhs) || isNan(rhs)) return false;
#endif

#ifdef USE_SIGNCHECK
    // Consider \a lhs and \a rhs unequal when they have a different
    // sign, even if they both are very close to zero.
    // \note The check for equality is needed because zero and negative zero
    // have different signs but are equal to each other.
    if (isNegative(lhs) != isNegative(rhs)) return lhs == rhs;
#endif

    floatUnion mlhs = { lhs };
    floatUnion mrhs = { rhs };
    floatMask_t ilhs = mlhs.mask;
    floatMask_t irhs = mrhs.mask;

    // Make \a ilhs and \a irhs lexicographically ordered as twos-complement
    // long.
    if (ilhs < 0) ilhs = floatNegativeMask - ilhs;
    if (irhs < 0) irhs = floatNegativeMask - irhs;

    // If \a ilhs and \a irhs are less than \a maxUlps apart, then \a lhs and
    // \a rhs are considered equal.
    return std::abs(ilhs - irhs) <= maxUlps;

  }


  bool Numeric::compare(double lhs, double rhs, doubleMask_t maxUlps)
  {

#ifdef USE_INFINITYCHECK
    // If either \a lhs or \a rhs is infinite, then return true if they're
    // exactly equal, i.e. they have the same sign.
    if (isInf(lhs) || isInf(rhs)) return lhs == rhs;
#endif

#ifdef USE_NANCHECK
    // Consider \a lhs and \a rhs unequal when either is NaN. 
    if (isNan(lhs) || isNan(rhs)) return false;
#endif

#ifdef USE_SIGNCHECK
    // Consider \a lhs and \a rhs unequal when they have a different
    // sign, even if they both are very close to zero.
    // \note The check for equality is needed because zero and negative zero
    // have different signs but are equal to each other.
    if (isNegative(lhs) != isNegative(rhs)) return lhs == rhs;
#endif

    doubleUnion mlhs = { lhs };
    doubleUnion mrhs = { rhs };
    doubleMask_t ilhs = mlhs.mask;
    doubleMask_t irhs = mrhs.mask;

    // Make \a ilhs and \a irhs lexicographically ordered as twos-complement
    // long.
    if (ilhs < 0) ilhs = doubleNegativeMask - ilhs;
    if (irhs < 0) irhs = doubleNegativeMask - irhs;

    // If \a ilhs and \a irhs are less than \a maxUlps apart, then \a lhs and
    // \a rhs are considered equal.
    ilhs -= irhs;
    if (ilhs < 0) ilhs = -ilhs;
    return ilhs <= maxUlps;

  }
  

} // namespace LOFAR
