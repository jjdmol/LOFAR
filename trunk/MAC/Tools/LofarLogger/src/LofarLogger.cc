//#  LofarLogger.cc: logger wrapper
//#
//#  Copyright (C) 2002-2003
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

#include "LofarLogger.h"
#include <stdarg.h>
#include <stdio.h>
#include <log4cplus/configurator.h>
#include <log4cplus/loggingmacros.h>

// namespaces
using namespace std;

// static member initialization
LofarLogger LofarLogger::_sLofarLogger;

/**
 * LofarLogger constructor
 */
LofarLogger::LofarLogger() :
    _loggerMap()
{
    log4cplus::PropertyConfigurator::doConfigure("log4cplus.properties");
}

LofarLogger::LofarLogger(const LofarLogger& rhs) :
    _loggerMap(rhs._loggerMap)
{
}

/**
 * ~LofarLogger destructor
 */
LofarLogger::~LofarLogger()
{
}

/**
 * getInstance - returns the static instance of the LofarLogger
 */
LofarLogger LofarLogger::getInstance()
{
  return _sLofarLogger;
}

/**
 * trace - logs TRACE level messages
 */
void LofarLogger::trace(const string& logger,const string& logEvent,const char* file, int line)
{
    log4cplus::Logger log4cplusLogger=getLogger(logger);

    if(log4cplusLogger.isEnabledFor(log4cplus::TRACE_LOG_LEVEL))
    {
        log4cplus::tostringstream _log4cplus_buf;
        _log4cplus_buf << logEvent;
        log4cplusLogger.forcedLog(log4cplus::TRACE_LOG_LEVEL,_log4cplus_buf.str(),file,line);
    }
}

/**
 * debug - logs DEBUG level messages
 */
void LofarLogger::debug(const string& logger,const string& logEvent,const char* file, int line)
{
    log4cplus::Logger log4cplusLogger=getLogger(logger);

    if(log4cplusLogger.isEnabledFor(log4cplus::DEBUG_LOG_LEVEL))
    {
        log4cplus::tostringstream _log4cplus_buf;
        _log4cplus_buf << logEvent;
        log4cplusLogger.forcedLog(log4cplus::DEBUG_LOG_LEVEL, _log4cplus_buf.str(), file,line);
    }
}

/**
 * info - logs INFO level messages
 */
void LofarLogger::info(const string& logger,const string& logEvent,const char* file, int line)
{
    log4cplus::Logger log4cplusLogger=getLogger(logger);

    if(log4cplusLogger.isEnabledFor(log4cplus::INFO_LOG_LEVEL))
    {
        log4cplus::tostringstream _log4cplus_buf;
        _log4cplus_buf << logEvent;
        log4cplusLogger.forcedLog(log4cplus::INFO_LOG_LEVEL, _log4cplus_buf.str(), file,line);
    }
}

/**
 * warn - logs WARN level messages
 */
void LofarLogger::warn(const string& logger,const string& logEvent,const char* file, int line)
{
    log4cplus::Logger log4cplusLogger=getLogger(logger);

    if(log4cplusLogger.isEnabledFor(log4cplus::WARN_LOG_LEVEL))
    {
        log4cplus::tostringstream _log4cplus_buf;
        _log4cplus_buf << logEvent;
        log4cplusLogger.forcedLog(log4cplus::WARN_LOG_LEVEL, _log4cplus_buf.str(), file,line);
    }
}

/**
 * error - logs ERROR level messages
 */
void LofarLogger::error(const string& logger,const string& logEvent,const char* file, int line)
{
    log4cplus::Logger log4cplusLogger=getLogger(logger);

    if(log4cplusLogger.isEnabledFor(log4cplus::ERROR_LOG_LEVEL))
    {
        log4cplus::tostringstream _log4cplus_buf;
        _log4cplus_buf << logEvent;
        log4cplusLogger.forcedLog(log4cplus::ERROR_LOG_LEVEL, _log4cplus_buf.str(), file,line);
    }
}

/**
 * fatal - logs FATAL level messages
 */
void LofarLogger::fatal(const string& logger,const string& logEvent,const char* file, int line)
{
    log4cplus::Logger log4cplusLogger=getLogger(logger);

    if(log4cplusLogger.isEnabledFor(log4cplus::FATAL_LOG_LEVEL)) 
    {
        log4cplus::tostringstream _log4cplus_buf;
        _log4cplus_buf << logEvent;
        log4cplusLogger.forcedLog(log4cplus::FATAL_LOG_LEVEL, _log4cplus_buf.str(), file,line);
    }
}

/**
 * formatString - creates a formatted string
 */
string LofarLogger::formatString(char* format,...)
{
    va_list args;
    char buffer[1000];
    va_start(args,format);
    vsnprintf(buffer,1000,format,args);
    return string(buffer);
}

/**
 * getLogger - finds a logger in the logger map or creates a new logger
 */
log4cplus::Logger& LofarLogger::getLogger(const string& logger)
{
    TLoggerMap::iterator loggerMapIterator=_loggerMap.find(logger);
    if(loggerMapIterator==_loggerMap.end())
    {
        _loggerMap.insert(TLoggerMap::value_type(logger,log4cplus::Logger::getInstance(logger)));
    }
    loggerMapIterator=_loggerMap.find(logger);
    return (loggerMapIterator->second);
}
