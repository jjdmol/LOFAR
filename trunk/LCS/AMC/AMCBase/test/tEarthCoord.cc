//#  tEarthCoord.cc: test program for the EarthCoord class
//#
//#  Copyright (C) 2002-2004
//#  ASTRON (Netherlands Foundation for Research in Astronomy)
//#  P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//#  This program is free software; you can redistribute it and/or modify
//#  it under the terms of the GNU General Public License as published by
//#  the Free Software Foundation; either version 2 of the License, or
//#  (at your option) any later version.
//#
//#  This program is distributed in the hope that it will be useful,
//#  but WITHOUT ANY WARRANTY; without even the implied warranty of
//#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//#  GNU General Public License for more details.
//#
//#  You should have received a copy of the GNU General Public License
//#  along with this program; if not, write to the Free Software
//#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//#  $Id$

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

//# Includes
#include <AMCBase/EarthCoord.h>
#include <Common/LofarLogger.h>
#include <Common/lofar_iostream.h>
#include <Common/Numeric.h>
#include <cmath>

using namespace LOFAR;
using namespace LOFAR::AMC;

int main(int, const char* argv[])
{
  INIT_LOGGER(argv[0]);

  try {

    EarthCoord ec0;
    EarthCoord ec1(0.25*M_PI, -0.33*M_PI, 1);
    EarthCoord ec2(-0.67*M_PI, 0.75*M_PI, 249.98, EarthCoord::WGS84);
    EarthCoord ec3(0.5*M_PI, 0.2*M_PI, -115.11, 
                   static_cast<EarthCoord::Types>(1294));
    vector<double> p;

    p = ec0.xyz();
    ASSERT(ec0.isValid() &&
           ec0.longitude() == 0 && 
           ec0.latitude() == 0 && 
           ec0.height() == 0 &&
           ec0.type() == EarthCoord::ITRF);
    ASSERT(p[0] == 0 &&
           p[1] == 0 &&
           p[2] == 0);

    p = ec1.xyz();
    ASSERT(ec1.isValid() &&
           ec1.longitude() == 0.25*M_PI && 
           ec1.latitude() == -0.33*M_PI &&
           ec1.height() == 1 &&
           ec1.type() == EarthCoord::ITRF);
    ASSERT(Numeric::compare(p[0],  0.3599466369818882) &&
           Numeric::compare(p[1],  0.3599466369818882) &&
           Numeric::compare(p[2], -0.8607420270039436));

    p = ec2.xyz();
    ASSERT(ec2.isValid() &&
           ec2.longitude() == -0.67*M_PI && 
           ec2.latitude() == 0.75*M_PI &&
           ec2.height() == 249.98 && 
           ec2.type() == EarthCoord::WGS84);

    ASSERT(Numeric::compare(p[0],  89.9794603127324) &&
           Numeric::compare(p[1], 152.1469583062028) &&
           Numeric::compare(p[2], 176.7625531610132));

    p = ec3.xyz();
    ASSERT(!ec3.isValid());

    cout << "ec0 = " << ec0 << endl;
    cout << "ec1 = " << ec1 << endl;
    cout << "ec2 = " << ec2 << endl;
    cout << "ec3 = " << ec3 << endl;

  }

  catch (Exception& e) {
    LOG_ERROR_STR(e);
    return 1;
  }

  return 2;  // forces assay to flag missing .stdout file as an error
}
