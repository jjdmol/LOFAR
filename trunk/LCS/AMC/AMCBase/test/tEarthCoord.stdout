ec0 = [0, 0, 0] (ITRF)
ec1 = [0.785398, -1.03673, 1] (ITRF)
ec2 = [-2.10487, 2.35619, 249.98] (WGS84)
ec3 = <INVALID>
